File: core/src/com/jupiter/europa/entity/ability/AttackAbility.java
package com.jupiter.europa.entity.ability;

import com.badlogic.ashley.core.Entity;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.jupiter.europa.EuropaGame;
import com.jupiter.europa.io.FileLocations;

import java.util.List;

/**
 * Created by nathan on 5/22/15.
 */
public class AttackAbility implements Ability {

    // Constants
    private static final String NAME = "Attack";
    private static final String ICON_NAME = "attack";
    private static final Cost COST = Cost.NONE;


    // Fields
    private final Entity entity;

    private TextureRegion sprite;


    // Initialization
    public AttackAbility(Entity entity) {
        this.entity = entity;
    }


    // Public Methods
    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public Cost getCost() {
        return COST;
    }

    @Override
    public Action getAction() {
        return null;
    }

    @Override
    public AbilityCategory getCategory() {
        return BasicAbilityCategories.ALL_ABILITIES;
    }

    @Override
    public TextureRegion getIcon() {
        if (this.sprite == null) {
            this.sprite = EuropaGame.game.getAssetManager().get(FileLocations.ICON_ATLAS, TextureAtlas.class).findRegion(ICON_NAME);
        }
        return this.sprite;
    }


    public static class AttackAction extends Action {

        //TODO: Basically Everything

        // Fields
        private final Entity entity;


        // Initialization
        private AttackAction(Entity entity) {
            this.entity = entity;
        }


        // Public Methods
        @Override
        public List<TargetInfo> getTargetRequirements() {
            return null;
        }

        @Override
        public void apply(List<Target> targets) {

        }
    }
}

File: core/src/com/jupiter/europa/entity/ability/AttackAbility.kt
/*
 * The MIT License
 *
 * Copyright 2015 Nathan Templon.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */

package com.jupiter.europa.entity.ability

import com.badlogic.ashley.core.Entity
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.graphics.g2d.TextureRegion
import com.jupiter.europa.EuropaGame
import com.jupiter.europa.io.FileLocations
import kotlin.properties.Delegates

/**
 * Created by nathan on 5/22/15.
 */
public class AttackAbility(private val entity: Entity) : Ability {

    private var sprite: TextureRegion = TextureRegion()


    // Public Methods
    override val name: String = NAME
    override val description: String = "Perform a basic attack with a melee or ranged weapon."
    override val cost: Cost = COST
    override val action: Action = AttackAction(entity)

    override val category: AbilityCategory = BasicAbilityCategories.ALL_ABILITIES

    override val icon: TextureRegion = ICON

    public class AttackAction internal constructor(entity: Entity) : Action(entity) {
    }

    companion object {

        // Constants
        private val NAME = "Attack"
        private val ICON_NAME = "attack"
        private val COST = Cost.NONE

        private val ICON by Delegates.lazy { EuropaGame.game.assetManager!!.get(FileLocations.ICON_ATLAS, javaClass<TextureAtlas>()).findRegion(ICON_NAME) }
    }
}
